### ISSUE TITLE
 [Runtime Bug] [Source Engine] NEOTOKYO° RTX-Remix broken

### ISSUE THREAD

**Issue by FrostKiwi** [NVIDIAGameWorks/rtx-remix#444]
**Describe the bug**
Game [NEOTOKYO°](https://store.steampowered.com/app/244630/NEOTOKYO/) fails to work with RTX Remix. It is a mod based on the Source Engine, so should work based on Half-Life² working. Specifically it is based on the Half-Life²: Episode 1 release version of the Source Engine, distributed via Steam's [Source SDK Base 2006](https://developer.valvesoftware.com/wiki/Source_SDK_Base)
![image](https://github.com/NVIDIAGameWorks/rtx-remix/assets/60887273/6e33ef3a-2cd8-4405-9bda-d7c70afdf8dd)

The game launches, the RTX-Remix Banner successfully appears and you can get into the RTX Remix menu. You can see which textures are loaded. Besides that, every single other feature is completely broken. Nothing works, it seems injecting into the rendering fails. `.USD` Scene capture also fails unfortunately, so the complete workflow is broken. The statistics display as shown in the screenshot below also shows all numbers being 0 for rendering time.

- GPU: RTX 4090
- Driver Version: 551.76
- Remix Version: 0.4.1
- Build info:
```
Protocol version 7
Exe version 1.0.0.1 (NeotokyoSource)
Exe build: 17:27:58 Dec  3 2009 (4044)
```

**Attach files!**
[hl2_d3d9.log](https://github.com/NVIDIAGameWorks/rtx-remix/files/14638365/hl2_d3d9.log)
[d3d9.log](https://github.com/NVIDIAGameWorks/rtx-remix/files/14638605/d3d9.log)

**Screenshots**
![image](https://github.com/NVIDIAGameWorks/rtx-remix/assets/60887273/32ae8dc1-1de8-4d09-b562-97d7284aa1d0)

**Motivation**
Why I'm interested in extending [NEOTOKYO°](https://store.steampowered.com/app/244630/NEOTOKYO/) via RTX-Remix: The game has some incredible art direction I fell in love with. I have exported the maps before into an external raytracer to get some images of the architecture on display. I would love to continue to do so with RTX-Remix.
![numa](https://github.com/NVIDIAGameWorks/rtx-remix/assets/60887273/1bfe207f-b95a-4f14-9611-7816e6fc9495)
![2](https://github.com/NVIDIAGameWorks/rtx-remix/assets/60887273/56f22f3f-0178-4409-a3a4-7f139041afda)

**Comment by anon-apple** [comment]
The Reflex Latency Stats display is broken due to a driver issue, nothing we can do about it. It has been reported though and has be fixed already so it will correct itself in a future driver version.

**Comment by FrostKiwi** [comment]
> The Reflex Latency Stats display is broken due to a driver issue

To clarify in regards to the compatibility as a whole: Due to a driver bug, the Stats display is broken. But even if that's fixed, the overall engine incompatibility will not be fixed, as it's a separate issue.

Is that assessment correct?

**Comment by anon-apple** [comment]
Yes, I don't know what's going on with the rest, if I had to guess it's just falling back to raster early for some reason but someone would have to look into it more. Was just commenting on the Latency Stats window because that's a known/fixed issue.

**Comment by AdamTadeusz** [comment]
@FrostKiwi If you're interested we've got a rebuild of the game in source 13 going, would love to see if the program works with the newer source engine version of the game. Here is a link to the repository https://github.com/NeotokyoRebuild/neo

**Comment by FrostKiwi** [comment]
The secret is running in dxlevel 7! `dxlevel 70` Now all features of RTX Remix work! On the flipside, this ruins a lot of the aspects of the engine, namely a hardcoded draw distance and complete lack of normal maps. Some shortcoming like forced low texture quality are fixable with `mat_picmap`, but other seem hardcoded.

After discovering this I found
https://developer.valvesoftware.com/wiki/DirectX_Versions
> DirectX 7.0 dxlevel 70 [RTX Remix](https://developer.valvesoftware.com/wiki/RTX_Remix) games ([Portal with RTX](https://developer.valvesoftware.com/wiki/Portal_with_RTX), [Half-Life 2 RTX](https://developer.valvesoftware.com/wiki/Half-Life_2_RTX)) runs on this DirectX level, then translated to Vulkan (using DXVK) and replacing assets at the same time, then renders as Direct3D 12 (DX12).
Not supported on [Source 2007](https://developer.valvesoftware.com/wiki/Orange_Box_(engine_branch)) and later (outside RTX Remix).

I gotta contact some of the devs who worked on this Portal with RTX I guess and ask how they managed to bend the source engine to work around these issues.

**Comment by FrostKiwi** [comment]
@anon-apple Now that I know how to get Remix working, do you happen to have the contact info of the devs who worked on Portal with RTX or Half life 2 RTX? I would love to figure out how they broke through the hardcoded limits of the source engine in `dxlevel 70` mode, like forced prop fadeout and no normal maps.

**Comment by anon-apple** [comment]
Well pretty much everyone working on Remix (myself included) worked on Portal RTX in some capacity (assuming they were around at the time), so depends on what you mean by that or what info you're looking for.

As for your question, we modified Source engine directly in various ways to accommodate Portal RTX, so some of the behavior you are seeing may be a result of that. I don't remember the full list of changes we did, usually just things that weren't possible to achieve purely with the options the engine exposes, for instance changes to culling logic. We did a fair few changes to the DirectX 7 rendering path though I know that much, mostly fixing some small issues with it necessary to get the data we needed passed to Remix, so that might have something to do with it.

It is possible though that things you're referring to specifically were simply done via some other methods though, if you look at Portal RTX's autoexec.cfg you'll find various tweaks in there to make it work and maybe one of those options is responsible for the prop fadeout behavior you mentioned. I do remember this being an issue but I don't remember precisely how it was fixed, as something like `cl_detailfade` I don't think had any effect on controlling it. As for the normal maps, we don't need normal maps from the base game if that's what you're referring to as all the material info in Portal RTX comes from replaced assets via Remix, so the only thing an object in the base game needs is an identifiable color texture to key off of.

**Comment by FrostKiwi** [comment]
Got the basic workflow working, as such I'm closing the issue.

# Source Engine 2006 / NEOTOKYO° and RTX Remix
- Rendering needs to run DirectX 7, the only way to make it be properly intercepted by trex pipeline. Launch with `dxlevel 70` or better `mat_dxlevel 70` in autoexec
- Source Engine 2006 only supports up to 3 worldlights. `r_worldlights 3` is the max, everything above that results in a crash
- Sky Rendering needs the Autodetect via Camera method, otherwise the sky renders above the scene, resulting in everything being broken
- Props in this mode will be forced out to fade early, this is called `Fade Scale`. This is a hardcoded feature and is described here: https://developer.valvesoftware.com/wiki/Prop_dynamic#Keyvalues
  - Explanation in Hammer:
  - > If you specify a fade in the worldspawn, or if the engine is running under dx7, then the engine will forcibly fade out props even if fademindist/fademaxdist isn't specified. This scale factor gives you some control over the fade. Using 0 here turns off the forcible fades. Numbers smaller than 1 cause the prop to fade out at further distances, and greater than 1 cause it to fade out at closer distances.
  - You need to disable it in Hammer for each prop, or patch props to have https://developer.valvesoftware.com/wiki/$noforcedfade
- Disable frustum culling and Occluder system as per the autoexec configs of HL2 Remix and Portal Remix
  - Either way, decompiling a map for editing in Hammer breaks prop Occluders, so they need to go either way.
  - Radiosity should also be disabled, as per HL2 Remix configs
- Textures are loaded at reduced size
  - In DirectX 7, this is hardcoded on level load and autoexec.cfg is not respected in this regard, in fact the whole MipMap system is disabled. But can luckily be overriden with `mat_picmip -10`, forcing all textures to their highest setting. This makes the USD capture include the full res textures as well
- Normal Maps and self-illumination maps aren't captured
  - I chased the idea of them being able to be captured *if only* we could enter DirectX 8. As explained to me on the RTX Remix discord and as I discovered myself: It doesn't matter, RTX Remix can't interpret or assign them correctly either way. Materials need to be rebuild from scratch in RTX Remix, which includes getting the self-illumination from the Base Color Texture's Alpha Channel, the Specular Map, which are separate files and converting them to PBR, and the normal map. RTX Remix wouldn't understand the correct assignments. The Workflow from there on out is fully manual.

There are more unknowns: Many mods rely on custom shaders, which aren't loaded due to being forced to DirectX 7. There is another Option described on the RTX Remix Discord, which involves creating `dxvk.conf` and inserting `d3d9.shaderModel=0`. This also makes RTX Remix work with the source engine and implies being able to launch in Direct X 9 mode, but from my understanding the source engine enters DirectX 7 Mode in this setup either way, so I don't think this method actually works. How replacement of these shaders will work in the context of RTX Remix, I'm not sure yet. Luckily, the basics work.

In general, you can't **just** simply replace stuff via the Marketed RTX Remix pipeline off --> Run Around Game --> Capture USD --> Replace with AI in RTX Remix --> Reinport USD for Asset Override. You need to setup Mapping tools, decompile the maps and rebuild with RTX Remix in mind, as a lot needs of engine rendering clashes with what RTX Remix needs. Ongoing point of research here as well.
